extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
                 continue;
             }
             BindableServiceBuildItem item = new BindableServiceBuildItem(service.name());
-            Set<MethodInfo> blockingMethods = gatherBlockingMethods(service);
-            for (MethodInfo method : blockingMethods) {
-                item.registerBlockingMethod(method.name());
+            Set<String> blockingMethods = gatherBlockingMethods(service, index);
+            for (String method : blockingMethods) {
+                item.registerBlockingMethod(method);
             }
             bindables.produce(item);
         }
     }
 
-    private Set<MethodInfo> gatherBlockingMethods(ClassInfo service) {
-        Set<MethodInfo> result = new HashSet<>();
-        boolean blockingClass = service.classAnnotation(GrpcDotNames.BLOCKING) != null;
-        if (blockingClass && service.classAnnotation(GrpcDotNames.NON_BLOCKING) != null) {
-            throw new DeploymentException("Class '" + service.name()
-                    + "' contains both @Blocking and @NonBlocking annotations.");
+    /**
+     * Generate list of {@link ClassInfo} with {@code service} as the first element and the class implementing
+     * {@code io.grpc.BindableService} (the protobuf generated {@code *ImplBase}) as the last one.
+     */
+    private static List<ClassInfo> classHierarchy(ClassInfo service, IndexView index) {
+        List<ClassInfo> collected = new ArrayList<>();
+        for (ClassInfo ci = service; ci != null;) {
+            collected.add(ci);
+
+            // Stop at the class that implements io.grpc.BindableService or
+            if (ci.interfaceNames().contains(GrpcDotNames.BINDABLE_SERVICE)) {
+                break;
+            }
+
+            DotName superName = ci.superName();
+            if (superName == null) {
+                break;
+            }
+            ci = index.getClassByName(superName);
         }
-        // if we have @Transactional, and no @NonBlocking,
-        blockingClass |= service.classAnnotation(TRANSACTIONAL) != null &&
-                service.classAnnotation(NON_BLOCKING) == null;
+        return collected;
+    }
 
-        for (MethodInfo method : service.methods()) {
-            if (BLOCKING_SKIPPED_METHODS.contains(method.name())) {
-                continue;
+    private enum BlockingMode {
+        UNDEFINED(false, false),
+        BLOCKING(true, true),
+        NON_BLOCKING(false, true),
+        TRANSACTIONAL(true, false);
```

Comment:
I wonder if we should use a different name; e.g. `IMPLICIT`, so that it's possible to reuse the constant for multiple annotations (and not only `@Transactional`).

Reply:
yup